runtime/src/accounts_db.rs

```diff
@@ -2150,27 +2762,231 @@ impl AccountsDB {
                 }
                 continue;
             }
-            for (offset, (_, account)) in rvs.iter().zip(&with_meta[infos.len()..]) {
-                storage.add_account();
+
+            for (offsets, (_, account)) in rvs
+                .windows(2)
+                .zip(&accounts_and_meta_to_store[infos.len()..])
+            {
+                let stored_size = offsets[1] - offsets[0];
+                storage.add_account(stored_size);
                 infos.push(AccountInfo {
                     store_id: storage.append_vec_id(),
-                    offset: *offset,
+                    offset: offsets[0],
+                    stored_size,
                     lamports: account.lamports,
                 });
             }
             // restore the state to available
             storage.set_status(AccountStorageStatus::Available);
         }
+
         self.stats
             .store_append_accounts
             .fetch_add(total_append_accounts_us, Ordering::Relaxed);
         self.stats
             .store_find_store
             .fetch_add(total_storage_find_us, Ordering::Relaxed);
-
         infos
     }
 
+    pub fn mark_slot_frozen(&self, slot: Slot) {
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            slot_cache.mark_slot_frozen();
+            slot_cache.report_slot_store_metrics();
+        }
+        self.accounts_cache.report_size();
+    }
+
+    // Force flush the cached roots, flush any unrooted frozen slots as well if there are
+    // > MAX_CACHE_SLOTS of them.
+    pub fn force_flush_accounts_cache(&self) {
+        self.flush_accounts_cache(true);
+    }
+
+    pub fn flush_accounts_cache_if_needed(&self) {
+        self.flush_accounts_cache(false);
+    }
+
+    fn flush_accounts_cache(&self, force_flush: bool) {
+        if !force_flush && self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS {
+            return;
+        }
+
+        // Flush all roots
+        let mut flush_roots_elapsed = Measure::start("flush_roots_elapsed");
+        let cached_roots = self.accounts_cache.clear_roots();
+        for root in &cached_roots {
+            self.flush_slot_cache(*root);
+            self.accounts_cache.set_max_flush_root(*root);
```

Comment:
considering heights of flushed roots are always increasing per `flush_accounts_cache` and the fn's call is serialized, maybe, we can make this `max_flush_root`  a local variable of this?

Oh, are you concerned about the case `cached_roots` are empty in a invocatoin?

Reply:
oh, maybe the hint is in this commit message? :) ef4d3bb4a259316a310c22b8eed913e8bd3817c5

Reply:
yeah, could be empty